### PR TITLE
Add a function for marking the alchemical ion

### PR DESCRIPTION
**Pull request by xiki-tempula**
Allow the alchemcial ion to be marked by

```
    ion = system[-1]
    ion = _mark_alchemical_ion(ion)
    system.updateMolecules(ion)
```

The following API could be supported
```
        * :meth:`~BioSimSpace.Sandpit.Exscientia._SireWrappers._system.System.getAlchemicalIon` to get the alchemical ion as a BSS molecule.
        * :meth:`~BioSimSpace.Sandpit.Exscientia._SireWrappers._system.System.getAlchemicalIonIdx` to get the index of alchemical ion as integer.
        * :meth:`~BioSimSpace.Sandpit.Exscientia._SireWrappers._molecule.Molecule.isAlchemicalIon` to check if a molecule is an alchemical ion return a bool.
```

### CODE REVIEW

File: python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    # Validate input.
+
+    if not isinstance(molecule, _Molecule):
+        raise TypeError(
+            "'molecule' must be of type 'BioSimSpace._SireWrappers.Molecule'"
+        )
+
+    # Cannot decouple a perturbable molecule.
+    if molecule.isAlchemicalIon():
+        raise _IncompatibleError(
```
Comment (msuruzhon) [inline python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py]: I am wondering if this should be raising an error or just a warning instead

File: python/BioSimSpace/Sandpit/Exscientia/_SireWrappers/_molecule.py [hunk 2]
```diff
@@ -518,6 +518,21 @@ def isML(self):
...
+        """
+        Whether this molecule is marked as Alchemical Ion.
+
+        Returns
+        -------
+
+        isAlchemicalIon : bool
+            Whether the molecule is marked as Alchemical Ion.
+        """
+        if self._sire_object.hasProperty("AlchemicalIon"):
```
Comment (msuruzhon) [inline python/BioSimSpace/Sandpit/Exscientia/_SireWrappers/_molecule.py]: `return self._sire_object.hasProperty("AlchemicalIon")`?

File: python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+from .._SireWrappers import Molecule as _Molecule
+from .._Exceptions import IncompatibleError as _IncompatibleError
+
+
+def _mark_alchemical_ion(molecule):
```
Comment (msuruzhon) [inline python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py:6]: Should this be private?

File: python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py (hunk 1 above)
Comment (xiki-tempula) [reply python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py]: This means that you are remarking an alchemical ion as an alchemical ion?
Currently this step is only executed in `fix_perturbed_charge` in BSS-wrappers so I think this should be a error?
I cannot think of a use case where we want to tolerate this thing?

File: python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py (hunk 3 above)
Comment (xiki-tempula) [reply python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py:6]: So in the later stage, this function will be executed as part of `fix_perturbed_charge`, which will be the public function for user to use.

File: python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py (hunk 1 above)
Comment (msuruzhon) [reply python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py]: I don't know, I am not sure idempotent operations should be raising an error. Imagine you are writing some code that makes sure some dodgy input is correctly marked. Then on top of calling this function you'd have to do error handling every single time... And I don't really see the benefit of that.

File: python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py (hunk 3 above)
Comment (msuruzhon) [reply python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py:6]: Ah so this will be in the public repo now?

Comment (xiki-tempula) [reply python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py:6]: Ben and Julien are happy with it so unless you don't want it to be public?

Comment (msuruzhon) [reply python/BioSimSpace/Sandpit/Exscientia/Align/_alch_ion.py:6]: No I am fine with that just wanted to make sure thanks.

**Review by msuruzhon — APPROVED** [review]

**Pull request merged**